src/ir/module-utils.h

```diff
@@ -392,29 +392,29 @@ template<typename T> struct CallGraphPropertyAnalysis {
   }
 };
 
-// Helper function for collecting the type signatures used in a module
+// Helper function for collecting all the types that are declared in a module,
+// which means the HeapTypes (that are non-basic, that is, not eqref etc., which
+// do not need to be defined).
 //
-// Used when emitting or printing a module to give signatures canonical
-// indices. Signatures are sorted in order of decreasing frequency to minize the
+// Used when emitting or printing a module to give HeapTypes canonical
+// indices. HeapTypes are sorted in order of decreasing frequency to minize the
 // size of their collective encoding. Both a vector mapping indices to
-// signatures and a map mapping signatures to indices are produced.
-inline void
-collectSignatures(Module& wasm,
-                  std::vector<Signature>& signatures,
-                  std::unordered_map<Signature, Index>& sigIndices) {
-  struct Counts : public std::unordered_map<Signature, size_t> {
-    void note(Signature sig) { (*this)[sig]++; }
+// HeapTypes and a map mapping HeapTypes to indices are produced.
+inline void collectHeapTypes(Module& wasm,
+                             std::vector<HeapType>& types,
+                             std::unordered_map<HeapType, Index>& typeIndices) {
+  struct Counts : public std::unordered_map<HeapType, size_t> {
+    bool isRelevant(Type type) { return !type.isBasic() && type.isRef(); }
+    void note(HeapType type) { (*this)[type]++; }
+    void note(Signature sig) { note(HeapType(sig)); }
```

Comment:
IIRC, `Signature` intentionally has an implicit conversion to `HeapType`, so we shouldn't need a separate overload for it.

Reply:
Oh, this compiles, but actually it's wrong I see... it will infinitely recurse without the explicit cast :rofl:

Reply:
Oh actually I misunderstood you, did you mean to remove the line entirely? Yes, that would both compile and be correct :smile: